internal/cmp/normalize.go

```diff
@@ -0,0 +1,104 @@
+package cmp
+
+import (
+	"cmp"
+	"fmt"
+	"math/rand"
+	"reflect"
+	"slices"
+	"sort"
+)
+
+type Normalizer[T any] interface {
+	Normalize() T
+}
+
+func SemanticEqual[T Normalizer[T]](this, that T) bool {
+	return reflect.DeepEqual(this.Normalize(), that.Normalize())
+}
+
+func NormalizeSlice[S ~[]E, E any](slice S, cmp func(a, b E) int) S {
+	if len(slice) == 0 {
+		return nil
+	}
+	slices.SortFunc(slice, cmp)
+	return slice
+}
+
+func Normalize(data any) error {
+	var err error
+	traverse(data, func(slice reflect.Value) {
+		sort.Slice(slice.Interface(), func(i, j int) bool {
+			iIface, jIface := slice.Index(i).Interface(), slice.Index(j).Interface()
+
+			if ok, result := compareSortable(iIface, jIface); ok {
+				return result < 0
+			}
+
+			result, e := ByJSON(iIface, jIface)
+			if e != nil {
+				err = fmt.Errorf("error converting slice %v to JSON: %w", slice, e)
+				return false
+			}
+			return result < 0
+		})
+	})
+	return err
+}
+
+func compareSortable(i, j any) (bool, int) {
+	iSortable, iSortableOK := i.(Sortable)
+	jSortable, jSortableOK := j.(Sortable)
+
+	if iSortableOK && jSortableOK {
+		return true, cmp.Compare(iSortable.Key(), jSortable.Key())
+	}
+
+	return false, -1
+}
+
+func PermuteOrder(data any, r *rand.Rand) {
+	traverse(data, func(slice reflect.Value) {
+		r.Shuffle(slice.Len(), func(i, j int) {
+			reflect.Swapper(slice.Interface())(i, j)
+		})
+	})
+}
```

Comment:
This one is only needed for test code right?

Reply:
potentially but similar as hashmaps in golang it could be a good practice to shuffle nested slices by default

Reply:
Yes, this one is only used for testing. I'll try to clear that in a separate PR.